### PR TITLE
[forward ad] Skip some metadata checks for 0 numel tensor

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81114
* __->__ #81055
* #79996

Fixes https://github.com/pytorch/pytorch/issues/80507

### CODE REVIEW

File: torch/csrc/autograd/autograd_meta.cpp [hunk 1]
```diff
@@ -88,7 +88,8 @@ bool has_same_meta(const Variable& base, const Variable& other) {
   if (!base.defined() || !other.defined()) {
     return false;
   }
-  if (base.storage_offset() != other.storage_offset()) {
+  if (base.storage_offset() != other.storage_offset() &&
+      !(base.numel() == 0 && other.numel() == 0)) {
```
Comment (ngimel) [inline torch/csrc/autograd/autograd_meta.cpp:92]:
```suggestion
if (base.numel() == 0 && other.numel() == 0) {
  return true; //0-element tensors are always fine
}
  if (base.storage_offset() != other.storage_offset()  {
```

Comment (soulitzer) [reply torch/csrc/autograd/autograd_meta.cpp]: Good catch. I also updated it so that the storage numel and conj check are reordered above.

**Review by ngimel — APPROVED** [review]

**Review by albanD — COMMENTED** [review]
From our discussion, I seem to remember that we wanted to have 3 cases here:
- Nothing needs to be done
- A view can be done
- Brand new memory needs to be allocated

Is it not the case anymore?

**Comment by soulitzer** [comment]
@albanD yeah see this older PR here https://github.com/pytorch/pytorch/pull/81049. The issue was that vmap(jvp layer norm started failing in core and functorch, and we'd like to avoid regressing that.
- The reason for the failure is that previously the pair of primal/tangent were considered to have the same metadata, but because we are more strict now about making sure the strides are exactly the same we are either going into _new_zeros_with_same_meta (if we decide to copy) or as_strided (if we decide to view) with a batched tensor. Both of these have a `checkBatchDimsAtFrontInLayout` check which leads to this failure.
- The new design in this PR is to more precisely define the invariant that we want to preserve in forward AD as "the same indices refer to the same relative location in storage", and any zero-numel tensor always satisfies that invariant, even if the stride is different. (See the code comment in this PR more details.) The main risk with this is if composite operations check the storage offset or stride properties to decide control flow, but that doesn't seem too composite compliant anyway.
- An alternative solution proposed in https://github.com/pytorch/pytorch/issues/80507#issuecomment-1178151979 is actually to relax the `checkBatchDimsAtFrontInLayout` check in _new_zeros_with_same_feature_meta or as_strided's batching rules.

**Pull request closed without merging**